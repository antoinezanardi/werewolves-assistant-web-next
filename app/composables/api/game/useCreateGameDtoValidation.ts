import { storeToRefs } from "pinia";
import { group as groupByField } from "radash";
import type { ComputedRef, Ref } from "vue";

import { MIN_PLAYERS_IN_GAME, MIN_PLAYERS_IN_GROUP } from "~/composables/api/game/constants/game.constants";
import type { CreateGamePlayerWithGroupDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import type { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { Role } from "~/composables/api/role/types/role.class";
import { useRolesStore } from "~/stores/role/useRolesStore";

type RoleWithMinInGame = Role & { minInGame: number };

type UseCreateGameDtoValidation = {
  isMinimumPlayersReached: ComputedRef<boolean>;
  areAllRolesSet: ComputedRef<boolean>;
  doesContainOneVillagerSidedRole: ComputedRef<boolean>;
  doesContainOneWerewolfSidedRole: ComputedRef<boolean>;
  areRolesMinimumPlayersReached: ComputedRef<boolean>;
  isTwoSistersRolePresentAndMinimumPlayersReached: ComputedRef<boolean>;
  isThreeBrothersRolePresentAndMinimumPlayersReached: ComputedRef<boolean>;
  areAdditionalCardsSetForThiefIfPresent: ComputedRef<boolean>;
  areAdditionalCardsSetForActorIfPresent: ComputedRef<boolean>;
  areAdditionalCardsSetForAdditionalCardsDependantRoles: ComputedRef<boolean>;
  arePlayerGroupsSetForPrejudicedManipulatorIfPresent: ComputedRef<boolean>;
  canCreateGame: ComputedRef<boolean>;
  gameCreationValidationErrors: ComputedRef<string[]>;
  isRolePresentAndMinimumPlayersReached: (role: RoleWithMinInGame) => boolean;
};

function useCreateGameDtoValidation(createGameDto: Ref<CreateGameDto>): UseCreateGameDtoValidation {
  const rolesStore = useRolesStore();
  const { roles } = storeToRefs(rolesStore);

  const { t } = useI18n();

  const isMinimumPlayersReached = computed<boolean>(() => createGameDto.value.players.length >= MIN_PLAYERS_IN_GAME);

  const areAllRolesSet = computed<boolean>(() => createGameDto.value.players.every(player => player.role.name !== undefined));

  const doesContainOneVillagerSidedRole = computed<boolean>(() => createGameDto.value.players.some(player => player.side.current === "villagers"));

  const doesContainOneWerewolfSidedRole = computed<boolean>(() => createGameDto.value.players.some(player => player.side.current === "werewolves"));

  const areRolesMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const rolesWithMinimumPlayers = roles.value.filter((role): role is RoleWithMinInGame => role.minInGame !== undefined);

    return rolesWithMinimumPlayers.every(role => isRolePresentAndMinimumPlayersReached(role));
  });

  const isTwoSistersRolePresentAndMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const role = roles.value.find(({ name }) => name === "two-sisters") as RoleWithMinInGame;

    return isRolePresentAndMinimumPlayersReached(role);
  });

  const isThreeBrothersRolePresentAndMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const role = roles.value.find(({ name }) => name === "three-brothers") as RoleWithMinInGame;

    return isRolePresentAndMinimumPlayersReached(role);
  });

  const areAdditionalCardsSetForThiefIfPresent = computed<boolean>(() => {
    const isThiefPresent = createGameDto.value.players.some(player => player.role.name === "thief");
    if (!isThiefPresent) {
      return true;
    }
    return createGameDto.value.additionalCards?.some(card => card.recipient === "thief") === true;
  });

  const areAdditionalCardsSetForActorIfPresent = computed<boolean>(() => {
    const isActorPresent = createGameDto.value.players.some(player => player.role.name === "actor");
    if (!isActorPresent) {
      return true;
    }
    return createGameDto.value.additionalCards?.some(card => card.recipient === "actor") === true;
  });

  const areAdditionalCardsSetForAdditionalCardsDependantRoles = computed<boolean>(() => areAdditionalCardsSetForActorIfPresent.value &&
    areAdditionalCardsSetForThiefIfPresent.value);

  const arePlayerGroupsSetForPrejudicedManipulatorIfPresent = computed<boolean>(() => {
    const isPrejudicedManipulatorPresent = createGameDto.value.players.some(player => player.role.name === "prejudiced-manipulator");
    if (!isPrejudicedManipulatorPresent) {
      return true;
    }
    const minimumGroups = 2;
    const playersWithGroup = createGameDto.value.players.filter(player => player.group !== undefined) as CreateGamePlayerWithGroupDto[];
    const groups = groupByField(playersWithGroup, player => player.group) as Record<string, CreateGamePlayerWithGroupDto[]>;
    const areAllGroupsMinimumPlayersReached = Object.values(groups).every(group => group.length >= MIN_PLAYERS_IN_GROUP);

    return Object.keys(groups).length >= minimumGroups && areAllGroupsMinimumPlayersReached;
  });

  const canCreateGame = computed<boolean>(() => isMinimumPlayersReached.value &&
    areAllRolesSet.value &&
    doesContainOneVillagerSidedRole.value &&
    doesContainOneWerewolfSidedRole.value &&
    areRolesMinimumPlayersReached.value &&
    areAdditionalCardsSetForThiefIfPresent.value &&
    areAdditionalCardsSetForActorIfPresent.value &&
    arePlayerGroupsSetForPrejudicedManipulatorIfPresent.value);

  const gameCreationValidationErrors = computed<string[]>(() => {
    const errors: Record<string, boolean> = {
      [t("composables.useCreateGameDtoValidation.minimumOfPlayersNotReached")]: !isMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.notAllRolesAreSet")]: !areAllRolesSet.value,
      [t("composables.useCreateGameDtoValidation.noVillagerSidedRole")]: !doesContainOneVillagerSidedRole.value,
      [t("composables.useCreateGameDtoValidation.noWerewolfSidedRole")]: !doesContainOneWerewolfSidedRole.value,
      [t("composables.useCreateGameDtoValidation.twoSistersMinimumPlayersNotReached")]: !isTwoSistersRolePresentAndMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.threeBrothersMinimumPlayersNotReached")]: !isThreeBrothersRolePresentAndMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.thiefAdditionalCardsNotSet")]: !areAdditionalCardsSetForThiefIfPresent.value,
      [t("composables.useCreateGameDtoValidation.actorAdditionalCardsNotSet")]: !areAdditionalCardsSetForActorIfPresent.value,
      [t("composables.useCreateGameDtoValidation.prejudicedManipulatorGroupsNotSet")]: !arePlayerGroupsSetForPrejudicedManipulatorIfPresent.value,
    };

    return Object.keys(errors).filter((key: keyof typeof errors) => errors[key]);
  });

  function isRolePresentAndMinimumPlayersReached(role: RoleWithMinInGame): boolean {
    const playersWithRole = createGameDto.value.players.filter(player => player.role.name === role.name);
    const playersWithRoleCount = playersWithRole.length;

    return playersWithRoleCount === 0 || playersWithRoleCount >= role.minInGame;
  }
  return {
    isMinimumPlayersReached,
    areAllRolesSet,
    doesContainOneVillagerSidedRole,
    doesContainOneWerewolfSidedRole,
    areRolesMinimumPlayersReached,
    isTwoSistersRolePresentAndMinimumPlayersReached,
    isThreeBrothersRolePresentAndMinimumPlayersReached,
    areAdditionalCardsSetForThiefIfPresent,
    areAdditionalCardsSetForActorIfPresent,
    areAdditionalCardsSetForAdditionalCardsDependantRoles,
    arePlayerGroupsSetForPrejudicedManipulatorIfPresent,
    canCreateGame,
    gameCreationValidationErrors,
    isRolePresentAndMinimumPlayersReached,
  };
}

export { useCreateGameDtoValidation };