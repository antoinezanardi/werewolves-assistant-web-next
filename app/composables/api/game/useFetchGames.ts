import type { CreateGameFeedbackDto } from "~/composables/api/game/dto/create-game-feedback/create-game-feedback.dto";
import type { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { Game } from "~/composables/api/game/types/game.class";
import { useWerewolvesAssistantApi } from "~/composables/api/useWerewolvesAssistantApi";

type UseFetchGames = {
  createGame: (createGameDto: CreateGameDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  getGame: (gameId: string) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  cancelGame: (gameId: string) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  makeGamePlay: (gameId: string, makeGamePlayDto: MakeGamePlayDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  createGameFeedback: (gameId: string, createGameFeedbackDto: CreateGameFeedbackDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
};

function useFetchGames(): UseFetchGames {
  const { fetchWerewolvesAssistantApi } = useWerewolvesAssistantApi();

  async function createGame(createGameDto: CreateGameDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games`, {
        method: "POST",
        body: JSON.stringify(createGameDto),
      });
    } catch {
      return null;
    }
  }

  async function getGame(gameId: string): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}`, { method: "GET" });
    } catch {
      return null;
    }
  }

  async function cancelGame(gameId: string): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}`, { method: "DELETE" });
    } catch {
      return null;
    }
  }

  async function makeGamePlay(gameId: string, makeGamePlayDto: MakeGamePlayDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}/play`, {
        method: "POST",
        body: JSON.stringify(makeGamePlayDto),
      });
    } catch {
      return null;
    }
  }

  async function createGameFeedback(gameId: string, createGameFeedbackDto: CreateGameFeedbackDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}/feedback`, {
        method: "POST",
        body: JSON.stringify(createGameFeedbackDto),
      });
    } catch {
      return null;
    }
  }
  return {
    createGame,
    getGame,
    cancelGame,
    makeGamePlay,
    createGameFeedback,
  };
}

export { useFetchGames };