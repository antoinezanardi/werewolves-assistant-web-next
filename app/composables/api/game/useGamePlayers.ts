import type { ComputedRef, Ref } from "vue";

import type { Game } from "~/composables/api/game/types/game.class";
import type { Player } from "~/composables/api/game/types/players/player.class";
import type { RoleName } from "~/composables/api/role/types/role.types";

type UseGamePlayers = {
  sheriffInPlayers: ComputedRef<Player | undefined>;
  getPlayersWithCurrentRole: (role: RoleName) => Player[];
};

function useGamePlayers(game: Ref<Game>): UseGamePlayers {
  const sheriffInPlayers = computed<Player | undefined>(() => game.value.players.find(({ attributes }) => attributes.some(({ name }) => name === "sheriff")));

  function getPlayersWithCurrentRole(role: RoleName): Player[] {
    return game.value.players.filter(player => player.role.current === role);
  }
  return {
    getPlayersWithCurrentRole,
    sheriffInPlayers,
  };
}

export { useGamePlayers };