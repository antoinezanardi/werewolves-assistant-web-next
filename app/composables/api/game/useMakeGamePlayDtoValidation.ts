import type { ComputedRef, Ref } from "vue";

import type { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { Game } from "~/composables/api/game/types/game.class";

type UseMakeGamePlayDtoValidation = {
  isCurrentGamePlayVoteTypeAndValid: ComputedRef<boolean>;
  isCurrentGamePlayTargetTypeAndValid: ComputedRef<boolean>;
  isCurrentGamePlayChooseCardTypeAndValid: ComputedRef<boolean>;
  isCurrentGamePlayChooseSideTypeAndValid: ComputedRef<boolean>;
  canCurrentPlayBeSkipped: ComputedRef<boolean>;
  canMakeGamePlay: ComputedRef<boolean>;
};

function useMakeGamePlayDtoValidation(makeGamePlayDto: Ref<MakeGamePlayDto>, game: Ref<Game>): UseMakeGamePlayDtoValidation {
  const isCurrentGamePlayVoteTypeAndValid = computed<boolean>(() => {
    if (game.value.currentPlay?.type !== "vote" || makeGamePlayDto.value.votes === undefined) {
      return false;
    }
    const voteInteraction = game.value.currentPlay.source.interactions?.[0];
    if (voteInteraction === undefined) {
      return false;
    }
    const { votes } = makeGamePlayDto.value;

    return votes.length >= voteInteraction.boundaries.min;
  });

  const isCurrentGamePlayTargetTypeAndValid = computed<boolean>(() => {
    if (game.value.currentPlay?.type !== "target" || makeGamePlayDto.value.targets === undefined) {
      return false;
    }

    const targetInteraction = game.value.currentPlay.source.interactions?.[0];
    if (targetInteraction === undefined) {
      return false;
    }
    const { targets } = makeGamePlayDto.value;

    return targets.length >= targetInteraction.boundaries.min;
  });

  const isCurrentGamePlayChooseCardTypeAndValid = computed<boolean>(() => game.value.currentPlay?.type === "choose-card" && makeGamePlayDto.value.chosenCardId !== undefined);

  const isCurrentGamePlayChooseSideTypeAndValid = computed<boolean>(() => game.value.currentPlay?.type === "choose-side" && makeGamePlayDto.value.chosenSide !== undefined);

  const canCurrentPlayBeSkipped = computed<boolean>(() => game.value.currentPlay?.canBeSkipped === true);

  const canMakeGamePlay = computed<boolean>(() => canCurrentPlayBeSkipped.value ||
    isCurrentGamePlayVoteTypeAndValid.value ||
    isCurrentGamePlayTargetTypeAndValid.value ||
    isCurrentGamePlayChooseCardTypeAndValid.value ||
    isCurrentGamePlayChooseSideTypeAndValid.value);

  return {
    isCurrentGamePlayVoteTypeAndValid,
    isCurrentGamePlayTargetTypeAndValid,
    isCurrentGamePlayChooseCardTypeAndValid,
    isCurrentGamePlayChooseSideTypeAndValid,
    canCurrentPlayBeSkipped,
    canMakeGamePlay,
  };
}

export { useMakeGamePlayDtoValidation };