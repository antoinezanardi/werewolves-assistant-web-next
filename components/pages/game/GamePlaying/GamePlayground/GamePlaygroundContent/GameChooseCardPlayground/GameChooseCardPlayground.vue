<template>
  <div id="game-choose-card-playground">
    GAME PLAYGROUND CHOOSE CARD
  </div>
</template>