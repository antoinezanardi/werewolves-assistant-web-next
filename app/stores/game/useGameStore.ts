import type { AsyncDataRequestStatus } from "nuxt/app";
import { defineStore } from "pinia";
import { unique } from "radash";

import type { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { GameOptions } from "~/composables/api/game/types/game-options/game-options.class";
import { Game } from "~/composables/api/game/types/game.class";
import { useFetchGames } from "~/composables/api/game/useFetchGames";
import { StoreIds } from "~/stores/enums/store.enum";
import { useGameEventsStore } from "~/stores/game/game-event/useGameEventsStore";

const useGameStore = defineStore(StoreIds.GAME, () => {
  const {
    getGame: fetchGameFromApi,
    cancelGame: cancelGameFromApi,
    makeGamePlay: makeGamePlayFromApi,
  } = useFetchGames();

  const game = ref<Game>(new Game());
  const gameOptions = computed<GameOptions>(() => game.value.options);

  const gamePlayerGroups = computed<string[]>(() => unique(game.value.players.map(player => player.group)).filter(Boolean) as string[]);

  const fetchingGameStatus = ref<AsyncDataRequestStatus>("idle");
  const cancelingGameStatus = ref<AsyncDataRequestStatus>("idle");
  const makingGamePlayStatus = ref<AsyncDataRequestStatus>("idle");

  const { resetGameEventIndex } = useGameEventsStore();

  function resetGame(): void {
    game.value = new Game();
  }

  async function fetchAndSetGame(gameId: string): Promise<void> {
    fetchingGameStatus.value = "pending";
    const fetchedGame = await fetchGameFromApi(gameId);
    if (!fetchedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    resetGameEventIndex();
    game.value = fetchedGame;
    fetchingGameStatus.value = "success";
  }

  async function cancelGame(): Promise<void> {
    cancelingGameStatus.value = "pending";
    const canceledGame = await cancelGameFromApi(game.value._id);
    if (!canceledGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    game.value = canceledGame;
    cancelingGameStatus.value = "success";
  }

  async function makeGamePlay(makeGamePlayDto: MakeGamePlayDto): Promise<void> {
    makingGamePlayStatus.value = "pending";
    const playedGame = await makeGamePlayFromApi(game.value._id, makeGamePlayDto);
    if (!playedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    resetGameEventIndex();
    game.value = playedGame;
    makingGamePlayStatus.value = "success";
  }

  async function skipGamePlay(): Promise<void> {
    return makeGamePlay({});
  }
  return {
    game,
    gameOptions,
    gamePlayerGroups,
    fetchingGameStatus,
    cancelingGameStatus,
    makingGamePlayStatus,
    resetGame,
    fetchAndSetGame,
    cancelGame,
    makeGamePlay,
    skipGamePlay,
  };
});

export { useGameStore };