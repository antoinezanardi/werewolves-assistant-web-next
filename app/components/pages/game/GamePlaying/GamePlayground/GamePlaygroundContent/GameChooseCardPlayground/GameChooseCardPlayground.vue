<template>
  <div id="game-choose-card-playground">
    <GlowCapture class="flex gap-2 h-full items-center justify-center w-full">
      <GameChooseCardPlaygroundAdditionalCard
        v-for="additionalCard in eligibleAdditionalCardsToChoose"
        :key="additionalCard._id"
        :additional-card="additionalCard"
        class="flex game-additional-card justify-center w-1/5"
        @click-additional-card="onClickFromAdditionalCardButton"
      />
    </GlowCapture>
  </div>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";
import GameChooseCardPlaygroundAdditionalCard from "~/components/pages/game/GamePlaying/GamePlayground/GamePlaygroundContent/GameChooseCardPlayground/GameChooseCardPlaygroundAdditionalCard/GameChooseCardPlaygroundAdditionalCard.vue";
import { useCurrentGamePlay } from "~/composables/api/game/game-play/useCurrentGamePlay";
import type { GameAdditionalCard } from "~/composables/api/game/types/game-additional-card/game-additional-card.class";
import { useMakeGamePlayDtoStore } from "~/stores/game/make-game-play-dto/useMakeGamePlayDtoStore";
import { useGameStore } from "~/stores/game/useGameStore";

const gameStore = useGameStore();
const { game } = storeToRefs(gameStore);

const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
const { makeGamePlayDto } = storeToRefs(makeGamePlayDtoStore);

const { getEligibleAdditionalCardsToChooseInCurrentGamePlay } = useCurrentGamePlay(game);

const eligibleAdditionalCardsToChoose = computed<GameAdditionalCard[]>(() => getEligibleAdditionalCardsToChooseInCurrentGamePlay());

function onClickFromAdditionalCardButton(additionalCard: GameAdditionalCard): void {
  if (makeGamePlayDto.value.chosenCardId === additionalCard._id) {
    makeGamePlayDto.value.chosenCardId = undefined;

    return;
  }
  makeGamePlayDto.value.chosenCardId = additionalCard._id;
}
</script>