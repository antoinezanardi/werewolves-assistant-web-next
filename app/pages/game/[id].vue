<template>
  <div
    id="game-page"
    class="container-fluid"
  >
    <div
      v-if="fetchingGameStatus === 'pending'"
      id="loading-game-container"
      class="flex flex-col h-full items-center justify-center"
    >
      <TextProgressSpinner
        id="loading-game-spinner"
        :text="$t('pages.game.loadingGame')"
      />
    </div>

    <GameNotFound
      v-else-if="fetchingGameStatus === 'error'"
      id="game-not-found"
      class="h-full"
    />

    <div
      v-else
      class="h-full"
    >
      <Transition
        mode="out-in"
        name="fade"
      >
        <Component
          :is="gameStatusComponentToRender"
          :key="game.status"
          class="h-full"
          @game-feedback-submitter-button-click="onClickFromGameFeedbackSubmitterButton"
        />
      </Transition>

      <GameFeedbackSubmitter ref="gameFeedbackSubmitter"/>
    </div>
  </div>
</template>

<script setup lang="ts">
import type { Component } from "vue";
import { storeToRefs } from "pinia";

import GameCanceled from "~/components/pages/game/GameCanceled/GameCanceled.vue";
import type { GameFeedbackSubmitterExposed } from "~/components/pages/game/GameFeedbackSubmitter/game-feedback-submitter.types";
import GameFeedbackSubmitter from "~/components/pages/game/GameFeedbackSubmitter/GameFeedbackSubmitter.vue";
import GameNotFound from "~/components/pages/game/GameNotFound/GameNotFound.vue";
import GameOver from "~/components/pages/game/GameOver/GameOver.vue";
import GamePlaying from "~/components/pages/game/GamePlaying/GamePlaying.vue";
import TextProgressSpinner from "~/components/shared/misc/TextProgressSpinner/TextProgressSpinner.vue";
import type { GameStatus } from "~/composables/api/game/types/game.types";
import { useAudioStore } from "~/stores/audio/useAudioStore";
import { useGameStore } from "~/stores/game/useGameStore";

const route = useRoute();
const gameStore = useGameStore();
const { fetchAndSetGame } = gameStore;
const { game, fetchingGameStatus } = storeToRefs(gameStore);

const gameFeedbackSubmitter = ref<GameFeedbackSubmitterExposed | null>(null);

const audioStore = useAudioStore();
const { loadAllAudios, fadeOutPlayingBackgroundAudio } = audioStore;

const { t } = useI18n();

useHead({
  title: t("pages.game.playingGame"),
  meta: [{ name: "robots", content: "noindex, nofollow" }],
});

const { id } = route.params;

const gameStatusComponentToRender = computed<Component>(() => {
  const gameStatusComponents: Record<GameStatus, Component> = {
    playing: GamePlaying,
    over: GameOver,
    canceled: GameCanceled,
  };

  return gameStatusComponents[game.value.status];
});

function onClickFromGameFeedbackSubmitterButton(): void {
  if (!gameFeedbackSubmitter.value) {
    throw createError("Game Feedback Submitter is not defined");
  }
  gameFeedbackSubmitter.value.showGameFeedbackSubmitter();
}

const gameId = Array.isArray(id) ? id[0] : id;
void fetchAndSetGame(gameId);
loadAllAudios();

onUnmounted(fadeOutPlayingBackgroundAudio);
</script>