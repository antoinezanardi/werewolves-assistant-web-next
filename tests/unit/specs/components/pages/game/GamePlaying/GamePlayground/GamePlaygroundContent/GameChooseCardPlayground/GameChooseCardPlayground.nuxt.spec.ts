import { createTestingPinia } from "@pinia/testing";
import type { mount } from "@vue/test-utils";

import GameChooseCardPlayground from "~/components/pages/game/GamePlaying/GamePlayground/GamePlaygroundContent/GameChooseCardPlayground/GameChooseCardPlayground.vue";
import { StoreIds } from "~/stores/enums/store.enum";
import { createFakeGame } from "@tests/unit/utils/factories/composables/api/game/game.factory";
import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";

describe("Game Choose Card Playground Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameChooseCardPlayground>>;
  const testingPinia = { initialState: { [StoreIds.GAME]: { game: createFakeGame() } } };

  async function mountGameChooseCardPlaygroundComponent(): Promise<ReturnType<typeof mount<typeof GameChooseCardPlayground>>> {
    return mountSuspendedComponent(GameChooseCardPlayground, { global: { plugins: [createTestingPinia(testingPinia)] } });
  }

  beforeEach(async() => {
    wrapper = await mountGameChooseCardPlaygroundComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });
});