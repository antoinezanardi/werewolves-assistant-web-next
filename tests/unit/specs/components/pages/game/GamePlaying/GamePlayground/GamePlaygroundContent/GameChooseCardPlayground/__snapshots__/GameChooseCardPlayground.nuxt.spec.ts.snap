// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Choose Card Playground Component > should match snapshot when rendered. 1`] = `
"<div id="game-choose-card-playground" class="flex gap-2 h-full items-center justify-center w-full">
  <game-choose-card-playground-additional-card-stub additionalcard="[object Object]" class="flex game-additional-card justify-center md:w-1/5 w-full"></game-choose-card-playground-additional-card-stub>
  <game-choose-card-playground-additional-card-stub additionalcard="[object Object]" class="flex game-additional-card justify-center md:w-1/5 w-full"></game-choose-card-playground-additional-card-stub>
  <game-choose-card-playground-additional-card-stub additionalcard="[object Object]" class="flex game-additional-card justify-center md:w-1/5 w-full"></game-choose-card-playground-additional-card-stub>
</div>"
`;

exports[`Game Choose Card Playground Component > should match snapshot without shallow rendering when render. 1`] = `
"<div id="game-choose-card-playground" class="flex gap-2 h-full items-center justify-center w-full">
  <div id="game-choose-card-playground-additional-card" class="flex game-additional-card justify-center md:w-1/5 w-full"><button id="additional-card-button" aria-label="components.GameChooseCardPlaygroundAdditionalCard.chooseCardForRecipient, {&quot;roleName&quot;:&quot;shared.role.name.seer&quot;,&quot;recipientDefiniteName&quot;:&quot;shared.role.definiteName.thief, 1&quot;}" class="border-4 border-transparent flex flex-col h-52 items-center me-2 p-3 rounded-lg" type="button" style="will-change: transform; transition: 300ms cubic-bezier(.03,.98,.52,.99); transform: perspective(1000px) rotateX(NaNdeg) rotateY(NaNdeg) scale3d(1, 1, 1);"><img class="w-[125px] h-[125px] border-4 border-gray-800 role-image rounded-lg h-[125px] w-[125px] mb-1 border-4 border-gray-800 role-image rounded-lg h-[125px] w-[125px] mb-1" width="125" height="125" alt="shared.role.name.seer" data-nuxt-img="" id="additional-card-image" src="/svg/misc/ripples.svg"><span id="additional-card-role-label" class="flex flex-col font-semibold grow justify-center">shared.role.name.seer</span>
      <div class="js-tilt-glare" style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; overflow: hidden;">
        <div class="js-tilt-glare-inner" style="position: absolute; top: 50%; left: 50%; transform: rotate(180deg) translate(-50%, -50%); opacity: 0; width: 0px; height: 0px; transition: opacity 300ms cubic-bezier(.03,.98,.52,.99);"></div>
      </div>
    </button></div>
  <div id="game-choose-card-playground-additional-card" class="flex game-additional-card justify-center md:w-1/5 w-full"><button id="additional-card-button" aria-label="components.GameChooseCardPlaygroundAdditionalCard.chooseCardForRecipient, {&quot;roleName&quot;:&quot;shared.role.name.werewolf&quot;,&quot;recipientDefiniteName&quot;:&quot;shared.role.definiteName.thief, 1&quot;}" class="border-4 border-transparent flex flex-col h-52 items-center me-2 p-3 rounded-lg" type="button" style="will-change: transform; transition: 300ms cubic-bezier(.03,.98,.52,.99); transform: perspective(1000px) rotateX(NaNdeg) rotateY(NaNdeg) scale3d(1, 1, 1);"><img class="w-[125px] h-[125px] border-4 border-gray-800 role-image rounded-lg h-[125px] w-[125px] mb-1 border-4 border-gray-800 role-image rounded-lg h-[125px] w-[125px] mb-1" width="125" height="125" alt="shared.role.name.werewolf" data-nuxt-img="" id="additional-card-image" src="/svg/misc/ripples.svg"><span id="additional-card-role-label" class="flex flex-col font-semibold grow justify-center">shared.role.name.werewolf</span>
      <div class="js-tilt-glare" style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; overflow: hidden;">
        <div class="js-tilt-glare-inner" style="position: absolute; top: 50%; left: 50%; transform: rotate(180deg) translate(-50%, -50%); opacity: 0; width: 0px; height: 0px; transition: opacity 300ms cubic-bezier(.03,.98,.52,.99);"></div>
      </div>
    </button></div>
  <div id="game-choose-card-playground-additional-card" class="flex game-additional-card justify-center md:w-1/5 w-full"><button id="additional-card-button" aria-label="components.GameChooseCardPlaygroundAdditionalCard.chooseCardForRecipient, {&quot;roleName&quot;:&quot;shared.role.name.hunter&quot;,&quot;recipientDefiniteName&quot;:&quot;shared.role.definiteName.thief, 1&quot;}" class="border-4 border-transparent flex flex-col h-52 items-center me-2 p-3 rounded-lg" type="button" style="will-change: transform; transition: 300ms cubic-bezier(.03,.98,.52,.99); transform: perspective(1000px) rotateX(NaNdeg) rotateY(NaNdeg) scale3d(1, 1, 1);"><img class="w-[125px] h-[125px] border-4 border-gray-800 role-image rounded-lg h-[125px] w-[125px] mb-1 border-4 border-gray-800 role-image rounded-lg h-[125px] w-[125px] mb-1" width="125" height="125" alt="shared.role.name.hunter" data-nuxt-img="" id="additional-card-image" src="/svg/misc/ripples.svg"><span id="additional-card-role-label" class="flex flex-col font-semibold grow justify-center">shared.role.name.hunter</span>
      <div class="js-tilt-glare" style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; overflow: hidden;">
        <div class="js-tilt-glare-inner" style="position: absolute; top: 50%; left: 50%; transform: rotate(180deg) translate(-50%, -50%); opacity: 0; width: 0px; height: 0px; transition: opacity 300ms cubic-bezier(.03,.98,.52,.99);"></div>
      </div>
    </button></div>
</div>"
`;
